rspace/src/main/scala/coop/rchain/rspace/history/RadixTree.scala

```diff
@@ -0,0 +1,783 @@
+package coop.rchain.rspace.history
+
+import cats.effect.Sync
+import cats.syntax.all._
+import cats.{Monad, Parallel}
+import scodec.bits.ByteVector
+
+import scala.annotation.tailrec
+import scala.collection.concurrent.TrieMap
+import scala.language.higherKinds
+
+/**
+  * Radix Tree implementation
+  */
+object RadixTree {
+
+  /**
+    * Node structure (EmptyItem, Leaf and NodePointer)
+    * There is only one type of element in Tree [[Node]] = Vector[Item]
+    */
+  sealed trait Item
+  final case object EmptyItem                                   extends Item
+  final case class Leaf(prefix: ByteVector, value: ByteVector)  extends Item
+  final case class NodePtr(prefix: ByteVector, ptr: ByteVector) extends Item
+
+  type Node = Vector[Item]
+
+  /**
+    * number of items always constant
+    */
+  val numItems = 256
+
+  /**
+    * Empty node consists only of [[EmptyItem]]'s
+    */
+  val emptyNode: Node = (0 until numItems).map(_ => EmptyItem).toVector
+
+  /**
+    * Binary codecs for serializing/deserializing Node in Radix tree
+    * Coding structure for items:
+    *   EmptyItem                   - Empty (not encode)
+
+    *   Leaf(prefix,value)    -> [item number] [second byte] [prefix0]..[prefixM] [value0]..[value31]
+    *                               where is: [second byte] -> bit7 = 0 (Leaf identifier)
+    *                                                          bit6..bit0 - prefix length = M (from 0 to 127)
+    *
+    *   NodePtr(prefix,ptr)   -> [item number] [second byte] [prefix0]..[prefixM] [ptr0]..[ptr31]
+    *                               where is: [second byte] -> bit7 = 1 (NodePtr identifier)
+    *                                                          bit6..bit0 - prefix length = M (from 0 to 127)
+    */
+  object codecs {
+    private val defSize = 32
+
+    def encode(node: Node): ByteVector = {
+      //calculating size of buffer
+      val sizeNode = node.size
+
+      @tailrec
+      def loopSize(index: Int, size: Int): Int =
+        if (sizeNode > index)
+          node(index) match {
+            case EmptyItem => loopSize(index + 1, size)
+
+            case Leaf(leafPrefix, value) =>
+              val sizePrefix = leafPrefix.size.toInt
+              assert(sizePrefix <= 127, "error during serialization (size of prefix more than 127)")
+              assert(
+                value.size == defSize,
+                "error during serialization (size of leafValue not equal 32)"
+              )
+              loopSize(index + 1, size + 2 + sizePrefix + defSize)
```

Comment:
Could you define the `2` here  as a constant like `defSize`?

Reply:
The `2` bytes are for `[item number] [second byte]`, right?

Reply:
Yes, I will do it.